rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodesRecoveryManager.java

```diff
@@ -75,8 +95,151 @@ public NodesLockRestorationManager apply(RMCore rmCore) {
         };
     }
 
-    public void restoreLocks(RMNode rmNode, Client caller) {
-        nodesLockRestorationManager.handle(rmNode, caller);
+    public void recoverNodeSourcesAndNodes() {
+        Collection<NodeSourceData> nodeSources = rmCore.getDbManager().getNodeSources();
+        logPersistedNodeSourceInfo(nodeSources);
+
+        for (NodeSourceData nodeSourceData : nodeSources) {
+            String nodeSourceName = nodeSourceData.getName();
+            if (NodeSource.DEFAULT_LOCAL_NODES_NODE_SOURCE_NAME.equals(nodeSourceName)) {
+                // will be recreated by SchedulerStarter
+                rmCore.getDbManager().removeNodeSource(nodeSourceName);
+            } else {
+                recoverDeployedNodeSourceIfNeeded(nodeSourceData, nodeSourceName);
+            }
+        }
+    }
+
+    private void recoverDeployedNodeSourceIfNeeded(NodeSourceData nodeSourceDescriptor, String nodeSourceName) {
+        try {
+            logger.info("Recover node source " + nodeSourceName);
+            rmCore.createNotActiveNodeSource(nodeSourceDescriptor);
+            if (nodeSourceDescriptor.getStatus().equals(NodeSourceStatus.NODES_DEPLOYED)) {
+                rmCore.deployNodeSource(nodeSourceName);
+            }
+        } catch (Throwable t) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1181)